core/include/kernel/interrupt.h

```diff
@@ -287,11 +319,137 @@ TEE_Result interrupt_alloc_add_handler(struct itr_chip *chip, size_t it_num,
 				       void *data,
 				       struct itr_handler **out_hdl);
 
+/*
+ * interrupt_alloc_add_main_handler() - Alloc/register an interrupt main handler
+ * @itr_num	Interrupt number
+ * @handler	Handler function
+ * @flags	Bitmask flag ITRF_*
+ * @itr_hdl	NULL or output pointer to allocated struct itr_handler
+ */
+static inline
+TEE_Result interrupt_alloc_add_main_handler(size_t it_num,
+					    itr_handler_t handler,
+					    uint32_t flags, void *data,
+					    struct itr_handler **out_hdl)
+{
+	return interrupt_alloc_add_handler(interrupt_main_get(), it_num,
+					   handler, flags, data, out_hdl);
+}
+
 /*
  * interrupt_remove_free_handler() - Remove/free a registered interrupt handler
  * @hdl		Interrupt handlers
  * This function is the counterpart of interrupt_alloc_add_handler().
  * This function may panic on non-NULL invalid @hdl reference.
  */
 void interrupt_remove_free_handler(struct itr_handler *hdl);
+
+/*
+ * dt_get_itr_func - Typedef of function to get an interrupt line in DT node
+ *
+ * @args	Reference to phandle arguments
+ * @data	Pointer to data given at clk_dt_register_clk_provider() call
+ * @res		Output result code of the operation:
+ *		TEE_SUCCESS in case of success
+ *		TEE_ERROR_DEFER_DRIVER_INIT if clock is not initialized
+ *		Any TEE_Result compliant code in case of error.
+ *
+ * Returns an allocated struct itr_line pointer referencing to an interrupt
+ * line or NULL if invalid description in which case @res provides the error
+ * code. The interrupt matches the reference found with properties of the
+ * pointer FDT node.
+ *
+ * Upon success, the interrupt is configured and consumer can add a handler
+ * function to the interrupt. Yet, the interrupt is not enabled until consumer
+ * calls interrupt_enable().
+ * Upon success, struct itr_line must point to memory allocated with malloc()
+ * or like as it is freed prior returning to caller function.
+ */
+typedef struct itr_line *(*dt_get_itr_func)(struct dt_pargs *args, void *data,
```

Comment:
I know we have a few cases where we return the error code like this, but they are not very nice. I would prefer if we returned the TEE_Result and instead had an `**out_itr` or such.

Reply:
This prototype is related to how type `get_of_device_func` was introduced with _dt_driver.h_, without a `TEE_Result` status:
https://github.com/OP-TEE/optee_os/blob/8c0c44c9d2f768a95beacd4605776cfe29c9bff6/core/include/kernel/dt_driver.h#L35-L36
which was introduced later as output argument:
https://github.com/OP-TEE/optee_os/blob/d8b14b46af9d2c7cee7f875e389e1a659bd123da/core/include/kernel/dt_driver.h#L41-L42

I admit there is not real need to use a `void *` return value and a `TEE_Result *` output value instead of the opposite. I can create a P-R to switch these back if you think it would be more consistent regarding the overall core implementation.

Reply:
Yes, please.

Reply:
Ok, i'll create a P-R for that.

Reply:
addressed by https://github.com/OP-TEE/optee_os/pull/6098.